controlplane/observability.go

```diff
@@ -60,67 +60,56 @@ func (o *ObservabilityService) SendMetric(ctx context.Context, req *connect.Requ
-			Name:       name,
-			DataPoint:  dal.MetricCounter{Value: dataPoint.GetAsInt()},
+func (o *ObservabilityService) extractSumMetrics(runnerKey model.RunnerKey, name string, metric *metricsv1.Metric_Sum) []dal.Metric {
+	return slices.Map(metric.Sum.DataPoints, func(dataPoint *metricsv1.NumberDataPoint) dal.Metric {
+		return newMetric(runnerKey, name, dataPoint, dal.MetricCounter{
+			Value: dataPoint.GetAsInt(),
 		})
-		if err != nil {
-			return errors.WithStack(err)
-		}
-	}
-	return nil
+	})
 }
 
-func (o *ObservabilityService) insertHistogramMetrics(ctx context.Context, runnerKey model.RunnerKey, name string, metric *metricsv1.Metric_Histogram) error {
-	for _, dataPoint := range metric.Histogram.DataPoints {
-		err := o.dal.InsertMetricEntry(ctx, dal.Metric{
-			RunnerKey:  runnerKey,
-			StartTime:  time.Unix(0, int64(dataPoint.StartTimeUnixNano)),
-			EndTime:    time.Unix(0, int64(dataPoint.TimeUnixNano)),
-			SourceVerb: getVerbRef(dataPoint.Attributes, true),
-			DestVerb:   getVerbRef(dataPoint.Attributes, false),
-			Name:       name,
-			DataPoint: dal.MetricHistogram{
-				Count:  int64(dataPoint.Count),
-				Sum:    int64(*dataPoint.Sum),
-				Bucket: dataPoint.BucketCounts,
-			},
+func (o *ObservabilityService) extractHistogramMetrics(runnerKey model.RunnerKey, name string, metric *metricsv1.Metric_Histogram) []dal.Metric {
+	return slices.Map(metric.Histogram.DataPoints, func(dataPoint *metricsv1.HistogramDataPoint) dal.Metric {
+		return newMetric(runnerKey, name, dataPoint, dal.MetricHistogram{
+			Count:  int64(dataPoint.Count),
+			Sum:    int64(*dataPoint.Sum),
+			Bucket: dataPoint.BucketCounts,
 		})
-		if err != nil {
-			return errors.WithStack(err)
-		}
-	}
-	return nil
+	})
 }
 
-func getVerbRef(attributes []*commonv1.KeyValue, isSource bool) schema.VerbRef {
-	var name string
-	var module string
-
+func fillMetricFromAttributes(metric *dal.Metric, attributes []*commonv1.KeyValue) {
 	for _, attr := range attributes {
-		if attr.Key == observability.SourceVerbKey && isSource {
-			name = attr.Value.GetStringValue()
-		}
-		if attr.Key == observability.SourceModuleKey && isSource {
-			module = attr.Value.GetStringValue()
-		}
-		if attr.Key == observability.DestVerbKey && !isSource {
-			name = attr.Value.GetStringValue()
-		}
-		if attr.Key == observability.DestModuleKey && !isSource {
-			module = attr.Value.GetStringValue()
+		switch attr.Key {
+		case observability.SourceVerbKey:
+			metric.SourceVerb.Name = attr.Value.GetStringValue()
+		case observability.SourceModuleKey:
+			metric.SourceVerb.Module = attr.Value.GetStringValue()
+		case observability.DestVerbKey:
+			metric.DestVerb.Name = attr.Value.GetStringValue()
+		case observability.DestModuleKey:
+			metric.DestVerb.Module = attr.Value.GetStringValue()
+		case observability.RequestIDKey:
+			metric.RequestID = attr.Value.GetIntValue()
 		}
 	}
+}
+
+// Interface common to Sum and Histogram
+type metricInterface interface {
```

Comment:
This refactoring removed quite a bit of duplicate code by using one of Go's superpowers - declaring a local interface with a narrow set of methods only used by this implementation. In this case, both counters and histogram protobuf types have these getters that we can use to deduplicate the boilerplate.